src/types/query.d.ts

```diff
@@ -66,20 +154,48 @@ export module query {
   export function IsCredentials(expr: ExprArg): Expr
   export function IsRole(expr: ExprArg): Expr
 
-  export function Get(ref: ExprArg, ts?: ExprArg): Expr
+  export function Get<T extends object>(
+    ref: Expr.Ref<T>,
+    ts?: ExprVal<number | Expr.Time>
+  ): Expr<T extends (infer U)[] ? U : T>
+
+  // TODO
   export function KeyFromSecret(secret: ExprArg): Expr
   export function Reduce(
     lambda: ExprArg,
     initial: ExprArg,
     collection: ExprArg
   ): Expr
-  export function Paginate(set: ExprArg, opts?: object): Expr
-  export function Exists(ref: ExprArg, ts?: ExprArg): Expr
 
-  export function Create(collection_ref: ExprArg, params?: ExprArg): Expr
+  export function Paginate<T extends object>(
+    set: Expr.SetRef<T>,
+    params?: ExprVal<PaginateParams>
+  ): Page<Expr.Ref<T>>
+
+  export function Exists(
+    ref: Expr.Ref<any>,
+    ts?: ExprVal<number>
+  ): Expr<boolean>
+
+  export function Create<T extends object>(
+    collection_ref: Expr.CollectionRef<T>,
+    params: ExprVal<CreateParams<T>>
+  ): Document<T>
+
+  // TODO
   export function Update(ref: ExprArg, params: ExprArg): Expr
   export function Replace(ref: ExprArg, params: ExprArg): Expr
-  export function Delete(ref: ExprArg): Expr
+
+  export function Delete(
+    ref: Expr.Ref<any>
+  ): Expr<{
+    database: string
```

Comment:
I don't get why the output of `Delete` would always have `database`field, this only happens if you are deleing a database

Reply:
good to know! 👍 

i'm not using the return value of `Delete` atm, so someone else will need to make the change